java/ql/lib/semmle/code/java/frameworks/MyBatis.qll

```diff
@@ -102,3 +104,95 @@ class MyBatisSqlOperationAnnotationMethod extends Method {
 class TypeParam extends Interface {
   TypeParam() { this.hasQualifiedName("org.apache.ibatis.annotations", "Param") }
 }
+
+module ProviderInjection {
+  private class MyBatisAbstractSQL extends RefType {
+    MyBatisAbstractSQL() { this.hasQualifiedName("org.apache.ibatis.jdbc", "AbstractSQL") }
+  }
+
+  private class MyBatisProvider extends RefType {
+    MyBatisProvider() {
+      this.hasQualifiedName("org.apache.ibatis.annotations",
+        ["Select", "Delete", "Insert", "Update"] + "Provider")
+    }
+  }
+
+  private class MyBatisAbstractSQLMethodNames extends string {
+    MyBatisAbstractSQLMethodNames() {
+      this in [
+          "SELECT", "OFFSET_ROWS", "FETCH_FIRST_ROWS_ONLY", "OFFSET", "LIMIT", "ORDER_BY", "HAVING",
+          "GROUP_BY", "WHERE", "OUTER_JOIN", "RIGHT_OUTER_JOIN", "LEFT_OUTER_JOIN", "INNER_JOIN",
+          "JOIN", "FROM", "DELETE_FROM", "SELECT_DISTINCT", "SELECT", "INTO_VALUES", "INTO_COLUMNS",
+          "VALUES", "INSERT_INTO", "SET", "UPDATE"
+        ]
+    }
+  }
+
+  class MyBatisInjectionSink extends DataFlow::Node {
+    MyBatisInjectionSink() {
+      exists(Annotation a, Method m, TypeLiteral type, Class c |
+        a.getType() instanceof MyBatisProvider and
+        type = a.getValue(["type", "value"]) and
+        c.hasMethod(m, type.getTypeName().getType()) and
+        m.hasName(a.getValue("method").(StringLiteral).getValue()) and
+        this.asExpr() = m.getBody().getAStmt().(ReturnStmt).getResult()
+      )
+    }
+  }
+
+  class MyBatisAdditionalTaintStep extends TaintTracking::AdditionalTaintStep {
```

Comment:
I think it's easier if you just remove this class and make `MyBatisProviderStep` and `MyBatisAbstractSQLToStringStep` extend `AdditionalTaintStep`. You don't need to add them to the `step` predicate of the specific configurations, since `AdditionalTaintStep` subclasses are applied globally (i.e. to all configurations).